lib/doorkeeper/oauth/discovery_response.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+module Doorkeeper
+  module OAuth
+    class DiscoveryResponse < BaseResponse
+      def initialize(root_url, url_builder)
+        @root_url = root_url
+        @url_builder = url_builder
+      end
+
+      def body
+        @body ||= {
+          issuer: issuer || @root_url,
+          authorization_endpoint: authorization_endpoint,
+          token_endpoint: token_endpoint,
+          revocation_endpoint: revocation_endpoint,
+          userinfo_endpoint: userinfo_endpoint,
+          scopes_supported: scopes_supported,
+          response_types_supported: response_types_supported,
+          response_modes_supported: response_modes_supported,
+          grant_types_supported: grant_types_supported,
+          token_endpoint_auth_methods_supported: token_endpoint_auth_methods_supported,
+          code_challenge_methods_supported: code_challenge_methods_supported,
+        }.merge(custom_discovery_data)
+      end
+
+      def status
+        :ok
+      end
+
+      def headers
+        {
+          "Cache-Control" => "public",
+          "Content-Type" => "application/json; charset=utf-8",
+        }
+      end
+
+      private
+
+      def config
+        @config ||= Doorkeeper.configuration
+      end
+
+      def url_for(**args)
+        @url_builder.call(**args)
+      end
+
+      def custom_discovery_data
+        config.custom_discovery_data.symbolize_keys
+      end
+
+      def issuer
+        config.issuer
+      end
+
+      def authorization_endpoint
+        mapping = Doorkeeper::Rails::Routes.mapping[:authorizations] || {}
+
+        url_for(
+          controller: mapping[:controllers] || "doorkeeper/authorizations",
+          action: 'new'
+        )
+      end
+
+      def token_endpoint
+        mapping = Doorkeeper::Rails::Routes.mapping[:tokens] || {}
+        
+        url_for(
+          controller: mapping[:controllers] || "doorkeeper/tokens",
+          action: 'create'
+        )
+      end
+
+      def userinfo_endpoint
+        nil
+      end
+
+      def revocation_endpoint
+        mapping = Doorkeeper::Rails::Routes.mapping[:tokens] || {}
+        
+        url_for(
+          controller: mapping[:controllers] || "doorkeeper/tokens",
+          action: 'revoke'
+        )
+      end
+
+      def scopes_supported
+        config.scopes.to_a
+      end
+
+      def response_types_supported
+        config.authorization_response_types
+      end
+
+      def response_modes_supported
+        config.authorization_response_flows.flat_map(&:response_mode_matches).uniq
+      end
+
+      def grant_types_supported
+        grant_types_supported = config.grant_flows.dup
+        grant_types_supported << 'refresh_token' if !!config.refresh_token_enabled?
```

Comment:
This is technically broken, because `use_refresh_token` configuration can be a block/proc, which accepts `server` and `context` which, well, I don't think we have here, nor would it be possible to have?

Reply:
Thinking about this more, if we did drastically change how refresh tokens work, as suggested in #1771, then we could have options like:

```
Doorkeeper.configure do
  orm :active_record

  # Enable using them in general
  use_refresh_tokens true

  # Set the expiry for refresh tokens:
  refresh_token_expires_in 30.days

  # Allow refresh tokens to only be used by specific clients:
  allow_refresh_tokens_for do |client, access_grant|
    access_grant.scopes.exists?('offline_access')
  end
end
```

Or something.